apps/studio/pages/project/[ref]/functions/new.tsx

```diff
@@ -0,0 +1,286 @@
+import { useOrgOptedIntoAi } from 'hooks/misc/useOrgOptedIntoAi'
+import { IS_PLATFORM } from 'lib/constants'
+import { useEdgeFunctionDeployMutation } from 'data/edge-functions/edge-functions-deploy-mutation'
+import { toast } from 'sonner'
+import { useRouter } from 'next/router'
+import { useParams } from 'common'
+import { useAppStateSnapshot } from 'state/app-state'
+import DefaultLayout from 'components/layouts/DefaultLayout'
+import EdgeFunctionsLayout from 'components/layouts/EdgeFunctionsLayout/EdgeFunctionsLayout'
+import { PageLayout } from 'components/layouts/PageLayout/PageLayout'
+import FileExplorerAndEditor from 'components/ui/FileExplorerAndEditor/FileExplorerAndEditor'
+import { EDGE_FUNCTION_TEMPLATES } from 'components/interfaces/Functions/Functions.templates'
+import {
+  Command_Shadcn_,
+  CommandEmpty_Shadcn_,
+  CommandGroup_Shadcn_,
+  CommandInput_Shadcn_,
+  CommandItem_Shadcn_,
+  CommandList_Shadcn_,
+  Popover_Shadcn_,
+  PopoverContent_Shadcn_,
+  PopoverTrigger_Shadcn_,
+} from 'ui'
+import { useFlag } from 'hooks/ui/useFlag'
+
+const NewFunctionPage = () => {
+  const router = useRouter()
+  const { ref } = useParams()
+  const project = useSelectedProject()
+  const isOptedInToAI = useOrgOptedIntoAi()
+  const includeSchemaMetadata = isOptedInToAI || !IS_PLATFORM
+  const { setAiAssistantPanel } = useAppStateSnapshot()
+  const edgeFunctionCreate = useFlag('edgeFunctionCreate')
+
+  // TODO (Saxon): Remove this once the flag is fully launched
+  useEffect(() => {
+    if (!edgeFunctionCreate) {
+      router.push(`/project/${ref}/functions`)
+    }
+  }, [edgeFunctionCreate, ref, router])
+
+  const [files, setFiles] = useState<
+    { id: number; name: string; content: string; selected?: boolean }[]
+  >([
+    {
+      id: 1,
+      name: 'index.ts',
+      selected: true,
+      content: `// Setup type definitions for built-in Supabase Runtime APIs
+import "jsr:@supabase/functions-js/edge-runtime.d.ts";
+
+console.info('server started');
+
+Deno.serve(async (req: Request) => {
+  const data = {
+    message: 'Hello from Supabase Edge Functions!',
+  };
+
+  return new Response(
+    JSON.stringify(data),
+    { headers: { 'Content-Type': 'application/json', 'Connection': 'keep-alive' }}
+  );
+});`,
+    },
+  ])
+  const [functionName, setFunctionName] = useState('')
+  const [open, setOpen] = useState(false)
+  const [isPreviewingTemplate, setIsPreviewingTemplate] = useState(false)
+  const [savedCode, setSavedCode] = useState<string>('')
+
+  const { mutateAsync: deployFunction, isLoading: isDeploying } = useEdgeFunctionDeployMutation({
+    onSuccess: () => {
+      toast.success('Successfully deployed edge function')
+      if (ref && functionName) {
+        router.push(`/project/${ref}/functions/${functionName}/details`)
+      }
+    },
+  })
+
+  const onDeploy = async () => {
+    if (isDeploying || !ref || !functionName) return
+
+    try {
+      await deployFunction({
```

Comment:
consistency: use `mutate` instead of `mutateAsync`, then add the onError handlers to the useEdgeFunctionDeployMutation. We typically only use `mutateAsync` when we're firing multiple mutations that need to wait on each other

the mutation also has its own error handler by default so the toast in the catch error is rendering an extra error toast
![image](https://github.com/user-attachments/assets/7eecb68e-c22f-4f8b-b1ea-9264014d47dc)

Reply:
^ on that note we probably need to add some client side validations for function names?
I ran into that error while setting the name as "Test Function" -> am guessing no spaces allowed